File: RPC/RPC_10_12_10_2024/Soluciones/ThirteenthAnnual_UCF_LocalProgrammingContest/local2024FinalRoundData/solutions/L_please/please.java

import java.util.*;

public class please {
   final static int CUTOFF = 1;
   final static int NO_CUTOFF = 2;

   public static class student implements Comparable<student> {
      int num, let, pl;
      public int compareTo(student o) {
         if (pl != o.pl) return Integer.compare(o.pl, pl);
         if (let != o.let) return Integer.compare(let, o.let);
         return Integer.compare(num, o.num);
      }
   }
   public static void printAnswer(int[][] g) {
      int gr = 100, let = 4;
      while (let != 0) {
         if (g[let][gr] != NO_CUTOFF) {
            System.out.print(gr);
            System.out.print((--let==0) ? "\n" : " ");
         }
         gr--;
      }
   }

   please() {
      Scanner sc = new Scanner(System.in);
      int n = sc.nextInt();
      student[] sts = new student[n];
      for (int i = 0; i < n; i++) {
         sts[i] = new student();
         sts[i].num = sc.nextInt();
         sts[i].let = 4 + 'A' - sc.next().charAt(0);
         if (sts[i].let == -1) sts[i].let = 0;
         sts[i].pl = sc.nextInt();
      }
      Arrays.sort(sts);
      int[][] tr = new int[5][101], val = new int[5][101];
      for (int i = 0; i < n; i++) {
         var s = sts[i];
         if (s.let <= s.num && 100 - s.num >= 4 - s.let) {
            for (int L = s.let; L < 5; L++) {
               for (int G = s.num; G < 101; G++) {
                  if (L == s.let && G == s.num) {
                     val[L][G]++;
                  } else if (G!=0) {
                     if (tr[L][G] == CUTOFF) {
                         val[L][G] = val[L-1][G-1];
                     } else if (tr[L][G] == NO_CUTOFF) {
                        val[L][G] = val[L][G-1];
                     } else if (L!=0) {
                        val[L][G] = Math.max(val[L][G-1], val[L-1][G-1]);
                     } else {
                        val[L][G] = val[L][G-1];
                     }
                  }
               }
            }
         }
         if (i == n-1 || s.pl != sts[i+1].pl) {
            for (int L = 0; L < 5; L++) {
               for (int G = 0; G < 101; G++) {
                  if (L!=0 && G!=0 && tr[L][G]==0 && val[L][G-1]!=val[L-1][G-1]) {
                     if (val[L][G-1] < val[L-1][G-1]) {
                        tr[L][G] = CUTOFF;
                     } else {
                        tr[L][G] = NO_CUTOFF;
                     }
                  }
               }
            }
            for (int L = 0; L < 5; L++) {
               for (int G = 0; G < 101; G++) {
                  val[L][G] = 0;
               }
            }
         }
      }
      printAnswer(tr);
   }

   public static void main(String[] Args) {
      new please();  
   }
}

File: RPC/RPC_10_12_10_2024/Soluciones/ThirteenthAnnual_UCF_LocalProgrammingContest/local2024FinalRoundData/solutions/L_please/please.cpp

#include <bits/stdc++.h>

#define CUTOFF 1
#define NO_CUTOFF 2

using namespace std;

using student = struct student;
using vs = vector<student>;
using vi = vector<int>;
using vvi = vector<vi>;


struct student {
   int num;
   int let;
   int pl;
   bool operator < (const student & o) {
      // Handle the highest pleases first!!!
      if (pl != o.pl)
         return pl > o.pl;
      if (let != o.let)
         return let < o.let;
      if (num != o.num);
         return num < o.num;
      return 0;
   }
};
   
void printAnswer(vvi & trans) {
   int gr = 100;
   int let = 4;
   while (let) {
      if (trans[let][gr] != NO_CUTOFF) {
         cout << gr;
         let--; 
         if (let) cout << " ";
         else cout << endl;
      }
      gr--;
   }
}


int main() {
   int n;
   cin >> n;

   vs sts(n);
   for (int i = 0; i < n; i++) {
      char let;
      cin >> sts[i].num >> let >> sts[i].pl;
      sts[i].let = 4 - (let - 'A');
      if (sts[i].let == -1) sts[i].let = 0;
   }
   sort(sts.begin(), sts.end());

   // 0 represents uncommited
   // 1 represents cutoff
   // 2 represents no cutoff
   vvi trans(5, vi(101));
   vvi val(5, vi(101));

   for (int ind = 0; ind < n; ind++) {
      auto s = sts[ind];
      // Remove impossible situtations
      if (!(s.let > s.num || s.let - 4 < s.num - 100)) {
          // Update the DP table
          for (int i = s.let; i < 5; i++) {
             for (int j = s.num; j < 101; j++) {
                if (i == s.let && j == s.num) {
                   val[i][j]++;
                   continue;
                }

                if (j) {
                   if (trans[i][j] == CUTOFF) {
                      val[i][j] = val[i-1][j-1];
                   } else if (trans[i][j] == NO_CUTOFF) {
                      val[i][j] = val[i][j-1];
                   } else if (i && val[i][j-1] <= val[i-1][j-1]) {
                      val[i][j] = val[i-1][j-1];
                   } else {
                      val[i][j] = val[i][j-1];
                   }
                }
             }
          }
      }

      if (ind == n-1 || sts[ind].pl != sts[ind+1].pl) {
         // TODO: Commit
         for (int i = 0; i < 5; i++) {
            for (int j = 0; j < 101; j++) {
               // Check for transition
               if (i && j && !trans[i][j] && val[i][j-1] != val[i-1][j-1]) {
                  if (val[i][j-1] < val[i-1][j-1]) {
                     trans[i][j] = CUTOFF;                     
                  } else {
                     trans[i][j] = NO_CUTOFF;
                  }
               }
               //cerr << " " << val[i][j];
            }
            //cerr << endl;
         }
         //cerr << endl;
         for (int i = 0; i < 5; i++)
             for (int j = 0; j < 101; j++)
               val[i][j] = 0;   
      }

   }
   printAnswer(trans);

   return 0;
}
